### ISSUE TITLE
[@vue/cli-plugin-typescript] Is there any way to make type checks work in .ts files in editor?

### ISSUE THREAD

**Issue by 07akioni** [vuejs/vue-cli#6173]
### What problem does this feature solve?
I find in vscode (with vetur), props check works in `.vue` file when editing.
![image](https://user-images.githubusercontent.com/18677354/103178241-09c6b100-48bc-11eb-8fa6-5352dab344bf.png)

However it doesn't work in `.ts` file.

Is there any way to make type checks work in .ts files?
![image](https://user-images.githubusercontent.com/18677354/103178256-21059e80-48bc-11eb-9a80-e5b37a73d10b.png)

### What does the proposed API look like?
Nil

**Comment by fangbinwei** [comment]
I can't reproduce this problem. My project is created by `@vue/cli@5.0.0-alpha.0`. Could you  provide the information in the issue bug template? It's a compile error or just a vetur error message ?

**Comment by 07akioni** [comment]
> @vue/cli@5.0.0-alpha.0

Type check simply not works in `main.ts`.

![image](https://user-images.githubusercontent.com/18677354/103193681-24c70e80-4918-11eb-8837-50f9ae750739.png)

I use `([Vue 3] typescript, eslint, unit-jest, e2e-cypress)` preset and only `main.ts` is modified as before.
![image](https://user-images.githubusercontent.com/18677354/103193655-15e05c00-4918-11eb-9fe4-75ebbbd206bf.png)

**Comment by 07akioni** [comment]
In `shims-vue.ts`:
```
/* eslint-disable */
declare module '*.vue' {
  import type { DefineComponent } from 'vue'
  const component: DefineComponent<{}, {}, any>
  export default component
}
```

I think it makes in `.ts` files, all the imported `.vue` things is not treated as its real type but a empty `defineComponent()`?

**Comment by fangbinwei** [comment]
you are right. As far as I know, you should define `declare module '@/components/HelloWorld.vue'` by yourself. Or use `HelloWorld.tsx`

**Comment by 07akioni** [comment]
> you are right. As far as I know, you should define `declare module '@/components/HelloWorld.vue'` by yourself. Or use `HelloWorld.tsx`. I'm not sure if `vetur` can support the feature you want

Vetur seems to make type checks in `.vue`'s script part work. However in `.ts` it doesn't work.

😭

**Comment by fangbinwei** [comment]
I think you need plugin like [vue-ts-plugin](https://github.com/HerringtonDarkholme/vue-ts-plugin), Unfortunately, it seems this plugin doesn't support vue3 now.

Related issue https://github.com/vuejs/vetur/issues/1322, https://github.com/vuejs/vetur/issues/673

**Comment by 07akioni** [comment]
> I think you need plugin like [vue-ts-plugin](https://github.com/HerringtonDarkholme/vue-ts-plugin), Unfortunately, it seems this plugin doesn't support vue3 now.
> 
> Related issue [vuejs/vetur#1322](https://github.com/vuejs/vetur/issues/1322), [vuejs/vetur#673](https://github.com/vuejs/vetur/issues/673), [vuejs/vetur#2016](https://github.com/vuejs/vetur/issues/2016)

Thanks.

It may take a long time to make vue3 development tool chains mature. 😄

**Comment by 07akioni** [comment]
https://github.com/znck/vue-developer-experience

It can make type check work.

![image](https://user-images.githubusercontent.com/18677354/103229658-77cbb080-496e-11eb-9298-a8ae8b1fda45.png)
![Uploading image.png…]()

**Comment by fangbinwei** [comment]
using this plugin, 
- I should run the Select TypeScript version command in VS Code to use the workspace version of TypeScript. It's not out of the box for vscode users if integrating this plugin in CLI. But this plugin can be mentioned in the documentation maybe.

- this plugin doesn't support `.vue` which doesn't contain script tag, we can use this plugin with `shims-vue.d.ts` together to solve this problem.

maybe it would be better to let `vetur` integrate it. https://github.com/vuejs/vetur/pull/2145. I'm not very sure now.

**Comment by 07akioni** [comment]
> using `@vuedx/typescript-plugin-vue`,
> 
> * I should run the Select TypeScript version command in VS Code to use the workspace version of TypeScript. It's not out of the box for vscode users if integrating this plugin in CLI. But this plugin can be mentioned in the documentation maybe.
> * this plugin doesn't support `.vue` which doesn't contain script tag, we can use this plugin with `shims-vue.d.ts` together to solve this problem.
> 
> maybe it would be better to let `vetur` integrate it. [vuejs/vetur#2145](https://github.com/vuejs/vetur/pull/2145). I'm not very sure now.

Currentl y

**Comment by 07akioni** [comment]
For people who wants to use typescript `tsc` to build their library, I build [`v2s`](https://github.com/07akioni/v2s)(vue to script) to help convert `.vue` file to `.ts` file and refactor all the related imports/exports.

It does the conversion in place. 

Firstly copy `src` to `copied-src`
```
copied-src/Button.vue (lang=ts)
copied-src/index.ts      (import Button from './Button.vue')
```

Run `npx v2s -dr copied-src/**/*` (`-dr` means delete source & refactor imports/exports)
```
copied-src/Button__render.ts
copied-src/Button__script.ts
copied-src/Button.ts
copied-src/index.ts      (import Button from './Button.ts')
```

Then you can use `tsc` to compile your library.
